src/sources/exec/mod.rs

```diff
@@ -358,23 +361,31 @@ async fn run_command(
 
     let pid = child.id();
 
-    spawn_reader_thread(
-        stdout_reader,
-        config.event_per_line,
-        config.maximum_buffer_size_bytes,
-        STDOUT,
-        sender,
-    );
-
-    while let Some((line, stream)) = receiver.recv().await {
-        let event = create_event(&config, &hostname, line, &Some(stream.to_string()), pid);
-
-        let _ = out
-            .send(event)
-            .await
-            .map_err(|_: crate::pipeline::ClosedError| {
-                error!(message = "Failed to forward events; downstream is closed.");
-            });
+    spawn_reader_thread(stdout_reader, decoder.clone(), STDOUT, sender);
+
+    let mut send_error = false;
+    while let (Some(((events, byte_size), stream)), false) = (receiver.recv().await, send_error) {
+        emit!(&ExecEventsReceived {
+            count: events.len(),
+            command: config.command_line().as_str(),
+            byte_size,
+        });
+
+        for mut event in events {
+            handle_event(
+                &config,
+                &hostname,
+                &Some(stream.to_string()),
+                pid,
+                &mut event,
+            );
+
+            if out.send(event).await.is_err() {
+                error!(message = "Failed to forward event; downstream is closed.");
+                send_error = true;
+                break;
```

Comment:
Fun fact: you could add a label the outer loop (e.g. `'outer: while ...`) and then `break 'outer` here instead of using the flag. Not sure if I'd actually recommend it, but maybe?

Reply:
Ah yeah, I like that! I think I have tried that before, but had some issues with labels on blocks being unstable or clippy complaining that a loop doesn't actually loop. Seems to work in this case, though